### ISSUE TITLE
how to get text feature encoded by PromptSRC text encoder?

### ISSUE THREAD

**Issue by lin-whale** [muzairkhattak/PromptSRC#10]
Hello, thanks for your amazing open-source work.
I'm using your code to finetune CLIP. After that, I want to get text feature output by the model directly. I'm confused about the code in TextEncoder, but tried following code. It seems that the input text does not affect the result. Please help me. Thanks.

```python
class promptSRC_custom_clip:
    def __init__(self) -> None:
        sys.path.insert(0, '/home/aistar/PromptSRC')
        global args
        from train import get_trainer
        self.trainer = get_trainer()
        self.trainer.load_model(args.finetune_clip_pretrained_dataset, epoch=50)
        self.trainer.model.cuda()
    def encode_image(self, image):
        custom_feat = self.trainer.model.image_encoder(image.type(self.trainer.model.dtype).cuda())
        return custom_feat
    def encode_text(self, text):
        custom_feat = self.trainer.model.text_encoder(self.trainer.model.prompt_learner(), text.type(self.trainer.model.dtype).cuda())
        return custom_feat
    def cuda(self):
        self.trainer.model.cuda()
    def eval(self):
        pass

model = promptSRC_custom_clip()
model.cuda()
model.eval()
```

```python
for one_prompt in prompt_list:
    text = torch.cat([clip.tokenize(one_prompt)])
    cal_similarity(text, model)

def cal_similarity(text, model):
    custom_feat = model.encode_text(text.cuda())
    custom_feat = torch.nn.functional.normalize(custom_feat, dim=-1)
    # [1000, 768] -> [1, 768]
    custom_feat = custom_feat[-1].unsqueeze(0)
    custom_feat = custom_feat.unsqueeze(0)
```

**Comment by SHIBOYA** [comment]
During the training process, the input file consists of label and text description. How is this done? Have some functions in dataloader been changed?

**Comment by muzairkhattak** [comment]
Hi @Hstwhale,

Thank you for showing interest in PromptSRC!

Regarding your query, kindly note that you will need to load the trained weights in the above-shown code. Right now, it seems that you are not loading the weights. Also, make sure that you are running the model in eval mode so that the model produces the text features trained with PromptSRC.

Lastly, you can [check this issue](https://github.com/muzairkhattak/multimodal-prompt-learning/issues/50#issue-2160704336) from our MaPLe repo, on how to instantiate a trained model.

Let me know if that is helpful. 
Kind regards!

**Comment by lin-whale** [comment]
> Hi @Hstwhale,
> 
> Thank you for showing interest in PromptSRC!
> 
> Regarding your query, kindly note that you will need to load the trained weights in the above-shown code. Right now, it seems that you are not loading the weights. Also, make sure that you are running the model in eval mode so that the model produces the text features trained with PromptSRC.
> 
> Lastly, you can [check this issue](https://github.com/muzairkhattak/multimodal-prompt-learning/issues/50#issue-2160704336) from our MaPLe repo, on how to instantiate a trained model.
> 
> Let me know if that is helpful. Kind regards!

@muzairkhattak 
Thanks for your reply!

I think I have loaded the weight, in the code```self.trainer.load_model(args.finetune_clip_pretrained_dataset, epoch=50)```. And the args.finetune_clip_pretrained_dataset I set to ```     finetune_clip_pretrained_dataset: Union[str, Path] = \
        "/home/aistar/data/promptSRC/output/imagenet_hospital/PromptSRC/vit_l14_c2_ep50_batch4_4+4ctx_few_shot_4shots/seed1"```.
And the code runs successfully. So I think the trained model has been loaded.

My problem is that if I have a text "word1 word2 word3 ...". How can I get the encoded text feature like "a photo of a apple" in promptSRC framework. I can't fully understand the code in file promptsrc.py. Can you give me some suggesstion?

Kind regards!

**Comment by lin-whale** [comment]
https://github.com/muzairkhattak/PromptSRC/blob/bb95c77b634d63488f2cad81ff4a72d53bdd06d5/trainers/promptsrc.py#L112
In the above line of code, the ```tokenized_prompts``` is of dimension (n_cls, n_tkn), but I don't understand why the following code calculates the largest token in a token sequence containing n_tkn tokens. What's the meaning of that?
https://github.com/muzairkhattak/PromptSRC/blob/bb95c77b634d63488f2cad81ff4a72d53bdd06d5/trainers/promptsrc.py#L67

**Comment by muzairkhattak** [comment]
Hi @Hstwhale,

Regarding your queries, kindly note the following:

> My problem is that if I have a text "word1 word2 word3 ...". How can I get the encoded text feature like "a photo of a apple" in promptSRC framework. I can't fully understand the code in file promptsrc.py. Can you give me some suggestion?

First of all, you should instantiate a model ([Here is how you can do](https://github.com/muzairkhattak/multimodal-prompt-learning/issues/50#issue-2160704336) it for MaPLe, you will need to change a bit for PromptSRC)

Afterward, you can  construct your text prompt, and pass it through the text encoder of PromptSRC as shown [in these lines.](https://github.com/muzairkhattak/PromptSRC/blob/bb95c77b634d63488f2cad81ff4a72d53bdd06d5/trainers/promptsrc.py#L191)


 

> In the above line of code, the tokenized_prompts is of dimension (n_cls, n_tkn), but I don't understand why the following code calculates the largest token in a token sequence containing n_tkn tokens. What's the meaning of that?

Kindly note that this is the standard way to obtain final text embedding in CLIP like models. For a given embedding sequence, we index the End of Statement token (EOS) which represents the complete text that is implemented in that line `x[torch.arange(x.shape[0]), tokenized_prompts.argmax(dim=-1)]`. So basically, `tokenized_prompts.argmax(dim=-1)` is the index of that EOS token. 

I hope that is helpful! 
Thank you!

**Comment by muzairkhattak** [comment]
Hi @lin-whale,

Could you tell us if your issue is resolved? I am closing the issue for now. 

Feel free to reopen it, or ask any questions incase there are any queries.

Thank you and kind regards!
